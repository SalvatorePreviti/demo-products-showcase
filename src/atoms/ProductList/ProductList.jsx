import React from 'react'
import PropTypes from 'prop-types'
import ProductItem from '../ProductItem/ProductItem'
import styles from './ProductList.scss'

function ProductList(props) {
  const { items, activeProductsById, onSelect } = props
  return (
    <div>
      {items.length ? (
        items.map(item => {
          const itemId = item.id
          return (
            <ProductItem
              key={itemId}
              itemId={itemId}
              title={item.title}
              active={activeProductsById[itemId] === true}
              description={item.description}
              onSelect={() => onSelect && onSelect(item)}
            />
          )
        })
      ) : (
        <div className={styles.empty}>No products found</div>
      )}
    </div>
  )
}

ProductList.propTypes = {
  items: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      description: PropTypes.string
    })
  ).isRequired,
  activeProductsById: PropTypes.object.isRequired,
  onSelect: PropTypes.func
}

export default ProductList
