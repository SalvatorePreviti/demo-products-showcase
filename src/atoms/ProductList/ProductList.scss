.empty {
  font-style: italic;
  font-size: smaller;
  color: #444;
}
