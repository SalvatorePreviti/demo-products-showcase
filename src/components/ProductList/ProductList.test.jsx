import React from 'react'
import { shallow } from 'enzyme'

import ProductList from './ProductList'

describe('ProductList', () => {
  describe('with a list of items, without selection', () => {
    const items = [
      { id: '1', title: 'title 1', description: 'desc 1', active: false },
      { id: '2', title: 'title 2', description: 'desc 2', active: false },
      { id: '3', title: 'title 3', description: 'desc 3', active: false },
      { id: '4', title: 'title 4', description: 'desc 4', active: false }
    ]

    function getComponent() {
      return <ProductList items={items} />
    }

    it('snapshots', () => {
      expect(shallow(getComponent())).toMatchSnapshot()
    })

    it('renders all elements', () => {
      const root = shallow(getComponent())
      const found = root.find('ProductItem')
      expect(found.map(x => x.props())).toMatchObject([
        { active: false, description: 'desc 1', title: 'title 1' },
        { active: false, description: 'desc 2', title: 'title 2' },
        { active: false, description: 'desc 3', title: 'title 3' },
        { active: false, description: 'desc 4', title: 'title 4' }
      ])
    })
  })

  describe('with a list of items, with some selections', () => {
    const items = [
      { id: '1', title: 'title 1', description: 'desc 1', active: false },
      { id: '2', title: 'title 2', description: 'desc 2', active: true },
      { id: '3', title: 'title 3', description: 'desc 3', active: false },
      { id: '4', title: 'title 4', description: 'desc 4', active: true }
    ]

    function getComponent() {
      return <ProductList items={items} />
    }

    it('snapshots', () => {
      expect(shallow(getComponent())).toMatchSnapshot()
    })

    it('renders all elements', () => {
      const root = shallow(getComponent())
      const found = root.find('ProductItem')
      expect(found.map(x => x.props())).toMatchObject([
        { active: false, description: 'desc 1', title: 'title 1' },
        { active: true, description: 'desc 2', title: 'title 2' },
        { active: false, description: 'desc 3', title: 'title 3' },
        { active: true, description: 'desc 4', title: 'title 4' }
      ])
    })
  })

  it('handles selection event', () => {
    const items = [
      { id: '1', title: 'title 1', description: 'desc 1', active: false },
      { id: '2', title: 'title 2', description: 'desc 2', active: false },
      { id: '3', title: 'title 3', description: 'desc 3', active: false },
      { id: '4', title: 'title 4', description: 'desc 4', active: false }
    ]

    let selected
    const root = shallow(<ProductList items={items} onSelect={x => (selected = x)} />)

    root
      .find('ProductItem')
      .at(2)
      .props()
      .onSelect(items[2])

    expect(selected).toBe(items[2])
  })
})
