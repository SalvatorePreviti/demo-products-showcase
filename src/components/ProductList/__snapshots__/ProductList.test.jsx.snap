// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList with a list of items, with some selections snapshots 1`] = `
<div>
  <ProductItem
    active={false}
    description="desc 1"
    itemId="1"
    key="1"
    title="title 1"
  />
  <ProductItem
    active={true}
    description="desc 2"
    itemId="2"
    key="2"
    title="title 2"
  />
  <ProductItem
    active={false}
    description="desc 3"
    itemId="3"
    key="3"
    title="title 3"
  />
  <ProductItem
    active={true}
    description="desc 4"
    itemId="4"
    key="4"
    title="title 4"
  />
</div>
`;

exports[`ProductList with a list of items, without selection snapshots 1`] = `
<div>
  <ProductItem
    active={false}
    description="desc 1"
    itemId="1"
    key="1"
    title="title 1"
  />
  <ProductItem
    active={false}
    description="desc 2"
    itemId="2"
    key="2"
    title="title 2"
  />
  <ProductItem
    active={false}
    description="desc 3"
    itemId="3"
    key="3"
    title="title 3"
  />
  <ProductItem
    active={false}
    description="desc 4"
    itemId="4"
    key="4"
    title="title 4"
  />
</div>
`;
